internal/metrics/metrics.go

```diff
@@ -0,0 +1,199 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"git.sr.ht/~spc/go-log"
+	"github.com/jakub-dzon/k4e-operator/models"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	"os"
+	"path"
+	"sync"
+	"time"
+)
+
+const (
+	// DefaultRetentionDuration specifies 7 days data retention
+	DefaultRetentionDuration = 7 * 24 * time.Hour
+	// DefaultMaxBytes specifies no block data size limit
+	DefaultMaxBytes = int64(0)
+)
+
+type DataPoint struct {
+	Time  int64
+	Value float64
+}
+type Series struct {
+	Labels     map[string]string
+	DataPoints []DataPoint
+}
+
+//go:generate mockgen -package=metrics -destination=storage_mock.go . Storage
+type Storage interface {
+	Deregister() error
+	GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error)
+	AddMetric(value float64, labelsMap map[string]string) error
+}
+
+type TSDB struct {
+	appliedOptions *tsdb.Options
+	db             *tsdb.DB
+	dbLock         sync.RWMutex
+}
+
+func NewTSDB(dataDir string) (*TSDB, error) {
+	metricsDir := path.Join(dataDir, "metrics")
+	if err := os.MkdirAll(metricsDir, 0755); err != nil {
+		return nil, fmt.Errorf("cannot create directory: %w", err)
+	}
+	opts := tsdb.DefaultOptions()
+
+	opts.RetentionDuration = int64(DefaultRetentionDuration / time.Millisecond)
+	opts.MaxBytes = DefaultMaxBytes
+
+	db, err := tsdb.Open(metricsDir, nil, nil, opts, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &TSDB{
+		db:             db,
+		dbLock:         sync.RWMutex{},
+		appliedOptions: opts,
+	}, err
+}
+
+func (t *TSDB) Deregister() error {
```

Comment:
The lock ensures that changes made by `Update` are visible to other go routines (which may be adding or reading metrics).

Reply:
The Deregister does not happen in Update, correct? I expect 'deregister' to wait for all routines that are using the metrics feature to stop (or at least stop using the feature). Otherwise, we close the DB while they are reading/writing.
In general, I would expect the deregister of the device to stop all routines and then do the termination/close actions.

Reply:
Yes, `Deregister` is not called in `Update`.
`Close` method, which is part of the `Deregister` operation waits for all the in-flight operations to end - it tries to take exclusive (write) lock which can only be taken if all read locks are released (taken by the other operations). 

As a side note - it actually doesn't matter if other goroutines complete their actions successfully - the metrics data is removed in the `Deregister` call.

Reply:
The exclusive lock makes the routines wait. But it does not make them stop using the TSDB object. Immediately after Deregister releases the lock these routines will access TSDB object. Now 'TSDB.db' is not nil (perhaps close should set it to nil). But can it be used after closing? without panic? even if we can use it without panic I believe we will see an error in the log. It's not a big problem but it's noise in the log.

Reply:
Having db nil would require additional nil checks when we use it. When TSDB is closed and someone tries to commit, simply an error is returned - no panics.